packages/hardhat/contracts/CGTSwap.sol

```diff
@@ -0,0 +1,158 @@
+        CGT = _CGT;
+        stable = _stable;
+    }
+
+    /// @notice Sets new value for margin in basis points
+    function setMargin(uint16 _margin) external onlyOwner {
+        margin = _margin;
+    }
+
+    /// @notice Sets new stable token
+    function setStable(IERC20 _stable) external onlyOwner {
+        stable = _stable;
+    }
+
+    /// @notice Pause token swaps
+    function pause() external onlyOwner {
+        _requireNotPaused();
+        _pause();
+    }
+
+    /// @notice Unpause token swaps
+    function unpause() external onlyOwner {
+        _requirePaused();
+        _unpause();
+    }
+
+    /**
+     * @notice Swap CGT for stablecoin
+     * @param amountIn input amount of CGT
+     */
+    function swap(uint256 amountIn) 
+        external 
+        whenNotPaused
+        nonReentrant
+    {
+        require(amountIn > 0, "Invalid amount");
+        _swap(amountIn);
+    }
+
+    /// @notice Transfers fee collected from the contract to the owner
+    function withdrawFeeCollected() external onlyOwner {
+        TransferHelper.safeTransfer(
+            address(stable), 
+            owner(),
+            totalFeeCollected
+        );
+        totalFeeCollected = 0;
+    }
+
+    /**
+     * @notice Transfers tokens from the contract to the owner
+     * @param token address of token to withdraw
+     * @param amount amount of tokens to withdraw
+     */
+    function withdrawTokens(address token, uint256 amount) external onlyOwner {
+        require(token != address(0), "Invalid token address");
+        TransferHelper.safeTransfer(
+            token, 
+            owner(),
+            amount
+        );
+    }
+
+    /**
+     * @notice Calculate estimated amount of stablecoin received
+     * @dev Uses Chainlink XAU price feed
+     * @param amountIn input amount of CGT
+     * @return estimated stablecoin amount out
+     */
+    function quoteStablecoinAmount(uint256 amountIn) public view returns (uint256) {
+        uint256 stableAmount = (amountIn * uint256(getLatestXAU_USDPrice()) * 3215075) / (10**8 * 10**8);
+        uint256 adjustedAmount = (stableAmount / DECIMAL_FACTOR) + round(stableAmount, DECIMAL_FACTOR);
+        uint256 netStablecoinAmount = (adjustedAmount * (10000 - margin)) / 10000;
+        return netStablecoinAmount;
+    }
+
+    /**
+     * @notice View stablecoin balance of the contract
+     * @return stablecoin balance
+     */
+    function getStablecoinBalance() public view returns (uint256) {
+        return stable.balanceOf(address(this));
+    }
+
+    /**
+     * @notice Returns the latest XAU-USD price
+     * @dev Calls XAU-USD data feed
+     * @return XAU-USD price
+     */
+    function getLatestXAU_USDPrice() public view returns (int256) {
+        (
+            /*uint80 roundID*/,
+            int price,
+            /*uint startedAt*/,
+            /*uint timeStamp*/,
+            /*uint80 answeredInRound*/
+        ) = priceFeed.latestRoundData();
+        return price;
```

Comment:
Here or elsewhere can we make sure the price is not too extreme, gold price unlikely will go above the range of $20-$2000 per gram in the near term.

Reply:
I have added a check for the same in the quoteStablecoinAmount() function.